tabular/src/autogluon/tabular/models/tab_transformer/tab_transformer_model.py

```diff
@@ -0,0 +1,469 @@
+""" TabTransformer model """
+import logging
+import os
+import time
+
+import numpy as np
+import pandas as pd
+from autogluon.core.utils.loaders import load_pkl
+from tqdm import tqdm
+
+from .hyperparameters.parameters import get_default_param
+from .hyperparameters.searchspaces import get_default_searchspace
+from ..abstract.abstract_model import AbstractModel
+from ...constants import BINARY, REGRESSION
+from autogluon.core.utils import try_import_torch
+
+logger = logging.getLogger(__name__)
+
+
+class TabTransformerModel(AbstractModel):
+    params_file_name = "tab_trans_params.pth"
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self.types_of_features = None
+        self.verbosity = None
+        self.temp_file_name = "tab_trans_temp.pth"
+        try_import_torch()
+
+    def _get_types_of_features(self, df):
+        """ Returns dict with keys: : 'continuous', 'skewed', 'onehot', 'embed', 'language', values = ordered list of feature-names falling into each category.
+            Each value is a list of feature-names corresponding to columns in original dataframe.
+            TODO: ensure features with zero variance have already been removed before this function is called.
+        """
+        if self.types_of_features is not None:
+            Warning("Attempting to _get_types_of_features for TabTransformerModel, but previously already did this.")
+
+        feature_types = self.feature_metadata.get_type_group_map_raw()
+
+        categorical_featnames = feature_types['category'] + feature_types['object'] + feature_types['bool']
+        continuous_featnames = feature_types['float'] + feature_types[
+            'int']  # + self.__get_feature_type_if_present('datetime')
+        language_featnames = []  # TODO: not implemented. This should fetch text features present in the data
+        valid_features = categorical_featnames + continuous_featnames + language_featnames
+
+        # TODO: Making an assumption that "feature_types_raw" above isn't used elsewhere, since feature_types_raw will
+        # still have features with periods (".") in them.
+        valid_features = [feat.replace(".", "/-#") for feat in valid_features]
+
+        if len(categorical_featnames) + len(continuous_featnames) + len(language_featnames) != df.shape[1]:
+            unknown_features = [feature for feature in df.columns if feature not in valid_features]
+
+            df = df.drop(columns=unknown_features)
+            self.features = list(df.columns)
+
+        self.types_of_features = []
+        for feature in valid_features:
+            if feature in categorical_featnames:
+                type = 'CATEGORICAL'
+            elif feature in continuous_featnames:
+                type = 'SCALAR'
+            elif feature in language_featnames:
+                type = 'TEXT'
+
+            self.types_of_features.append({"name": feature, "type": type})
```

Comment:
This method appears mostly copied from tabular-NN. To avoid duplicate code, consider removing the 2 functions and importing 1 function from elsewhere in both TT and tabular-NN.

Reply:
Added common method in AbstractModel that they're both using now.